07_VirtualFileSystem/02_VirtualFileSystem.md

```diff
@@ -0,0 +1,352 @@
+
+One last thing about the initialization, since our kernel is loaded fully in memory, we don't actually need to have a file system mounted for the kernel to run (even if probably in the future we will need one) so the initialization is just optional to be done during boot time, if the the os already has a kind of shell we can initialize it on the first mount when it will be called. 
+
+#### Finding the correct mountpoint
+
+Now that we know how to handle the mountpoints, we need to understand how given a path find the correct route toward the right mountpoint. Depending on the approach how a path is defined can vary slightly: 
+
+* if we are using a single root approach we will have a path in the form of: `/path/to/folder/and/file`
+* if we are using a multi-root approach the the path will be similar to: `<device_id>/path/to/folder/and/file` (where device_id is what identify the file system to be used, and the device, what it is depend on the os, it can be a letter, a number, a mix, etc.)
+
+We will cover the single root approach, but eventually changing to a multi-root approach should be pretty easy. One last thing to keep in mind is that the path separator is another design decision, mostly every operating system use either "/" or "\" (the latter is mostly on windows os and derivatives), but in theory everything can be used as a path separator, we will stick with the unix-friendly "/", just keep in mind if going for the "windows" way, the separator is the same as the escape character, so it can interfere with the escape sequences.
+
+For example let's assume that we have the following list of mountpoints: 
+
+```c
+mountpoints[0].mountpoint = "/"
+mountpoints[1].mountpoint = "/home/mount"
+mountpoints[2].mountpoint = "/usr"
+mountpoints[3].mountpoint = "/home"
+```
+
+And we want to access the following paths: 
+
+* /home/user/folder1/file
+* /home/mount/folder1/file
+* /opt
+
+As we can see the first two paths have a common part, but belongs to different file system so we need to implement a function that given a path return the index of the file system it belongs to. 
+
+How to do it is pretty simple, we scan the array, and search for the "longest" mountpoint that is contained in the path, so for in the first example, we can see that there are two items in the array that are contained in the path: "/" (0), and "/home" (3), and the longest one is number 3, so this is what our function is going to return. 
+
+The second path instead has three mountpoints contained into it: "/" (0), "/home/mount" (1), "/home", in this case we are going to return 1. 
+
+The last one, has only one mountpoint that is contained into the path, and it is "/" (0). 
+
+In a single root scenario, there is always at least a mountpoint that is contained into the given path, and this is the root folder "/".
+
+What if for example path doesn't start with a "/"? this means that it's a relative path, it will be explained soon.
```

Comment:
It might be nicer for the reader if we include a small sentence explaining how relative paths work, and then we can say we'll look at it more later. At some point a relative path has to become absolute, this can happen in the program calling the vfs itself (meaning the vfs only has to support absolute paths), or you could include a field in the thread control block that stores the current working directory, and use that.